File: day16_state/guru/cpp/state.cpp
#include <iostream>
#include <memory>

#include "context.hpp"
#include "state.hpp"


void State::SetContext(std::weak_ptr<Context> context) {
  this->context_ = context;
}

void ConcreteStateA::Handle1() {
  std::cout << "ConcreteStateA handles Request1." << '\n';
  std::cout << "ConcreteStateA wants to change the state of the context." << '\n';
  if (auto this_context = this->context_.lock()) {
    this_context->TransitionTo(std::make_unique<ConcreteStateB>());
  } else {
    std::cout << "Error: " << typeid(this->context_).name() << " expired in " << __func__ << '\n';
  }
}

void ConcreteStateA::Handle2() {
  std::cout << "ConcreteStateA handles Request2." << '\n';
}

void ConcreteStateB::Handle1() {
  std::cout << "ConcreteStateB handles Request1." << '\n';
}

void ConcreteStateB::Handle2() {
  std::cout << "ConcreteStateB handles Request2." << '\n';
  std::cout << "ConcreteStateB wants to change the state of the context." << '\n';
  if (auto this_context = this->context_.lock()) {
    this_context->TransitionTo(std::make_shared<ConcreteStateA>());
  } else {
    std::cout << "Error: " << typeid(this->context_).name() << " expired in " << __func__ << '\n';
  }
}

File: day16_state/guru/py/state.py
from __future__ import annotations
from abc import ABC, abstractmethod
from overrides import overrides

import context as c


class State(ABC):
    """
    The base State class declares methods that all Concrete State should
    implement and also provides a backreference to the Context object,
    associated with the State. This backreference can be used by States to
    transition the Context to another State.
    """

    @property
    def context(self) -> c.Context:
        return self._context

    @context.setter
    def context(self, context: c.Context) -> None:
        self._context = context

    @abstractmethod
    def handle_1(self) -> None:
        pass

    @abstractmethod
    def handle_2(self) -> None:
        pass


"""
Concrete States implement various behaviors, associated with a state of the
Context.
"""
class ConcreteStateA(State):
    @overrides
    def handle_1(self) -> None:
        print('ConcreteStateA handles request_1.')
        print('ConcreteStateA wants to change the state of the context.')
        self._context.transition_to(ConcreteStateB())

    @overrides
    def handle_2(self) -> None:
        print('ConcreteStateA handles request_2.')


class ConcreteStateB(State):
    @overrides
    def handle_1(self) -> None:
        print('ConcreteStateB handles request_1.')

    @overrides
    def handle_2(self) -> None:
        print('ConcreteStateB handles request_2.')
        print('ConcreteStateB wants to change the state of the context.')
        self._context.transition_to(ConcreteStateA())
